File: problems/0133/solution.cpp
#include <queue>
#include <unordered_map>
#include <vector>
using namespace std;

class Node {
 public:
  int val;
  vector<Node*> neighbors;
  Node() {
    val = 0;
    neighbors = vector<Node*>();
  }
  Node(int _val) {
    val = _val;
    neighbors = vector<Node*>();
  }
  Node(int _val, vector<Node*> _neighbors) {
    val = _val;
    neighbors = _neighbors;
  }
};

class Solution {
 public:
  Node* cloneGraph(Node* node) {
    if (node == nullptr) return nullptr;
    queue<Node*> q;
    unordered_map<Node*, Node*> mp;
    q.push(node);
    mp[node] = new Node(node->val);
    while (!q.empty()) {
      int size = q.size();
      for (int i = 0; i < size; i++) {
        Node* curr = q.front();
        q.pop();
        for (auto& n : curr->neighbors) {
          if (mp.find(n) == mp.end()) {
            mp[n] = new Node(n->val);
            q.push(n);
          }
          mp[curr]->neighbors.push_back(mp[n]);
        }
      }
    }
    return mp[node];
  }
};

File: problems/0133/solution.py
from collections import deque
from typing import Optional


class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors: list[Node] = neighbors if neighbors is not None else []


class Solution:
    def cloneGraph(self, node: Optional["Node"]) -> Optional["Node"]:
        if not node:
            return None
        q = deque([node])
        mp: dict[Node, Node] = {node: Node(node.val)}
        while q:
            size = len(q)
            for _ in range(size):
                curr = q.popleft()
                for n in curr.neighbors:
                    if n not in mp:
                        mp[n] = Node(n.val)
                        q.append(n)
                    mp[curr].neighbors.append(mp[n])
        return mp[node]
